_posts/2023-05-26-correlated-subqueries-in-sql.md

```diff
@@ -0,0 +1,294 @@
+
+```sql
+SELECT uniquecarrier, origincityname, destcityname, flightdate, distance
+FROM ontime AS ontime_outer
+WHERE NOT EXISTS (
+     SELECT uniquecarrier, origincityname, destcityname, flightdate, distance
+     FROM ontime
+     WHERE origin=ontime_outer.origin AND dest=ontime_outer.dest AND flightdate>ontime_outer.flightdate
+);
+```
+
+| uniquecarrier |           origincityname           |            destcityname            | flightdate | distance |
+|---------------|------------------------------------|------------------------------------|------------|----------|
+| AA            | Daytona Beach, FL                  | Charlotte, NC                      | 2017-02-27 | 416.0    |
+| EV            | Abilene, TX                        | Dallas/Fort Worth, TX              | 2017-02-15 | 158.0    ||
+| EV            | Dallas/Fort Worth, TX              | Durango, CO                        | 2017-02-13 | 674.0    |
+
+#### IN / ANY / ALL
+
+`IN` can be used to check if a *given value* exists within the result returned by the subquery. For example, we can obtain a list of all carriers that have performed more than `250 000` flights in the dataset using the following query:
+
+```sql
+SELECT uniquecarrier
+FROM ontime
+GROUP BY uniquecarrier
+HAVING COUNT(*) > 250000;
+```
+
+We can then use an `IN` clause to obtain all flights performed by those carriers.
+
+```sql
+SELECT *
+FROM ontime
+WHERE uniquecarrier IN (
+     SELECT uniquecarrier
+     FROM ontime
+     GROUP BY uniquecarrier
+     HAVING COUNT(*) > 250000
+);
+```
+
+A correlated subquery can be useful here if we want to not count the total amount of flights performed by each carrier, but count the total amount of flights *for the given route*. We can select all flights performed by carriers that have performed *at least 1000 flights on a given route* using the following query.
+
+```sql
+SELECT *
+FROM ontime AS ontime_outer
+WHERE uniquecarrier IN (
+     SELECT uniquecarrier
+     FROM ontime
+     WHERE ontime.origin=ontime_outer.origin AND ontime.dest=ontime_outer.dest
+     GROUP BY uniquecarrier
+     HAVING COUNT(*) > 1000
+);
+```
+
+`ANY` and `ALL` are generalizations of `IN`. `IN` checks if the value is present in the set returned by the subquery. This is equivalent to `= ANY(...)`. The `ANY` and `ALL` operators can be used to perform other comparison operators (such as `>`, `<`, `<>`). The above query can be rewritten to `ANY` in the following form.
+
+```sql
+SELECT *
+FROM ontime AS ontime_outer
+WHERE uniquecarrier = ANY (
+     SELECT uniquecarrier
+     FROM ontime
+     WHERE ontime.origin=ontime_outer.origin AND ontime.dest=ontime_outer.dest
+     GROUP BY uniquecarrier
+     HAVING COUNT(*) > 1000
+);
+```
+
+#### Performance
+
+Logically, correlated subqueries are executed *once per row*. As such, it is natural to think that correlated subqueries are very expensive and should be avoided for performance reasons.
```

Comment:
Maybe to make the difference between regular scalar subqueries and correlated subqueries more clear, we can add something along the line of:
"Whereas scalar subqueries are logically executed *once*, correlated subqueries are logically executed *once per row*. Slight modifications to a query that may seem harmless, such as adding a `WHERE` clause that compares a column from a subquery with a column from the outer query, may have significant performance implications."